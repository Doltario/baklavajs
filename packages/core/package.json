{
    "name": "@baklavajs/core",
    "version": "2.0.2-beta.3",
    "description": "Core logic of BaklavaJS, a graph editor / node editor for the web",
    "author": "newcat <freddy.wagner@web.de>",
    "homepage": "https://newcat.github.io/baklavajs",
    "license": "MIT",
    "files": [
        "dist/*"
    ],
    "main": "dist/cjs/index.js",
    "types": "dist/index.d.ts",
    "module": "dist/esm/index.js",
    "scripts": {
        "build:esm": "tsc -p tsconfig.build.json --declaration false --outDir dist/esm --module esnext",
        "build:cjs": "tsc -p tsconfig.build.json --declaration false --outDir dist/cjs",
        "build:declaration": "tsc -p tsconfig.build.json --emitDeclarationOnly --outDir dist",
        "build": "rimraf dist && yarn run build:esm && yarn run build:cjs && yarn run build:declaration",
        "test": "jest --coverage --colors"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/newcat/baklavajs.git"
    },
    "bugs": {
        "url": "https://github.com/newcat/baklavajs/issues"
    },
    "dependencies": {
        "@baklavajs/events": "^2.0.2-beta.3",
        "uuid": "^9.0.0"
    },
    "devDependencies": {
        "@types/jest": "^29.5.2",
        "@types/uuid": "^9.0.1",
        "jest": "^29.5.0",
        "rimraf": "^5.0.1",
        "ts-jest": "^29.1.0",
        "ts-node": "^10.9.1",
        "typedoc": "^0.24.7",
        "typescript": "^5.1.3"
    },
    "publishConfig": {
        "access": "public"
    },
    "gitHead": "723bf8b6249e6b67d9a7a5a9c84cf1ee0ca33739"
}
