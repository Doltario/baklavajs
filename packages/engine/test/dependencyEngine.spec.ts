import { Editor, NodeInterface, defineNode } from "@baklavajs/core";
import { TestNode } from "./testNode";
import {
    AfterNodeCalculationEventData,
    BeforeNodeCalculationEventData,
    DependencyEngine,
    allowMultipleConnections,
} from "../src";

describe("DependencyEngine", () => {
    it("emits the beforeNodeCalculation and afterNodeCalculation events", async () => {
        const editor = new Editor();
        const n1 = editor.graph.addNode(new TestNode())!;
        const n2 = editor.graph.addNode(new TestNode())!;
        editor.graph.addConnection(n1.outputs.c, n2.inputs.a);

        const engine = new DependencyEngine<void>(editor);
        const beforeSpy = jest.fn();
        const afterSpy = jest.fn();
        engine.events.beforeNodeCalculation.subscribe("a", beforeSpy);
        engine.events.afterNodeCalculation.subscribe("b", afterSpy);

        n1.inputs.a.value = 2;
        n1.inputs.b.value = 3;
        n2.inputs.b.value = 4;

        await engine.runOnce();

        expect(beforeSpy).toHaveBeenCalledTimes(2);
        expect(beforeSpy.mock.calls[0][0]).toEqual({
            node: n1,
            inputValues: {
                a: 2,
                b: 3,
            },
        } as BeforeNodeCalculationEventData);
        expect(beforeSpy.mock.calls[1][0]).toEqual({
            node: n2,
            inputValues: {
                a: 5,
                b: 4,
            },
        } as BeforeNodeCalculationEventData);

        expect(afterSpy).toHaveBeenCalledTimes(2);
        expect(afterSpy.mock.calls[0][0]).toEqual({
            node: n1,
            outputValues: {
                c: 5,
                d: -1,
            },
        } as AfterNodeCalculationEventData);
        expect(afterSpy.mock.calls[1][0]).toEqual({
            node: n2,
            outputValues: {
                c: 9,
                d: 1,
            },
        } as AfterNodeCalculationEventData);
    });

    it("handles nodes without a calculate method", async () => {
        const editor = new Editor();
        const NoCalculationNode = defineNode({ type: "NoCalculation" });
        editor.graph.addNode(new NoCalculationNode());
        const engine = new DependencyEngine<void>(editor);
        expect(await engine.runOnce()).toEqual(new Map());
    });

    it("allows using multiple connections", async () => {
        const editor = new Editor();
        const spy = jest.fn();
        const MultiNode = defineNode({
            type: "MultiNode",
            inputs: {
                a: () => new NodeInterface<number[]>("a", [0]).use(allowMultipleConnections),
            },
            calculate({ a }) {
                spy(a);
                return {};
            },
        });
        const n1 = editor.graph.addNode(new TestNode())!;
        const n2 = editor.graph.addNode(new MultiNode())!;
        editor.graph.addConnection(n1.outputs.c, n2.inputs.a);
        editor.graph.addConnection(n1.outputs.d, n2.inputs.a);

        const engine = new DependencyEngine<void>(editor);
        await engine.runOnce();

        expect(spy).toHaveBeenCalledWith([2, 0]);
    });

    it("handles nodes without calculate functions", async () => {
        const editor = new Editor();
        const NoCalculateNode = defineNode({
            type: "NoCalculateNode",
            outputs: {
                a: () => new NodeInterface("a", 3),
            }
        });
        const n1 = editor.graph.addNode(new NoCalculateNode())!;
        const n2 = editor.graph.addNode(new TestNode())!;
        editor.graph.addConnection(n1.outputs.a, n2.inputs.a);

        const engine = new DependencyEngine<void>(editor);
        const result = await engine.runOnce();

        expect(result).toMatchSnapshot();
    });
});
