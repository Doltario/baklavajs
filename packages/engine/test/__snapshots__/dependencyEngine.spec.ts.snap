// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependencyEngine handles nodes without calculate functions 1`] = `
Map {
  "3595507a-28e8-4dac-a536-82f1cfb9b7cf" => Map {
    "a" => 3,
  },
  "1ed1bcb1-6582-4606-9336-32c69330e8d1" => Map {
    "c" => 4,
    "d" => 2,
  },
}
`;
